toreview/BitSetUtils.java

```diff
@@ -0,0 +1,174 @@
+        if (numBits > 0 && numBits % minLength == 0) {
+            int numLong = numBits / minLength;
+            long[] numbers = new long[numLong];
+            for (int i = 0; i < numLong; i++) {
+                numbers[numLong - (i + 1)] = Double.doubleToLongBits(hash(toHash) / (i + 1));
+            }
+            return BitSet.valueOf(numbers);
+        } else throw new IllegalArgumentException(NOT_VALID_NUMBIT_EXCEPTION_MSG);
+    }
+
+    /**
+     * @param toHash  the String (a key) of which generate hashcode
+     * @param numBits number of bits that the hash code will be constituted of, must be > 0 &&  <= 160
+     * @return the bitSet containing the hash (SHA-1) of the bytes in input, truncated to numBits, bitSet's size is <= numBits
+     * @throws IllegalArgumentException if peer isn't valid or the numBit isn't in its bounds
+     */
+    public static BitSet hash(@NonNull byte[] toHash, int numBits) {
+        if (numBits > 0 && numBits <= 160) {
+            byte[] digest = {0};
+            try {
+                MessageDigest md = MessageDigest.getInstance(SHA_1);
+                md.update(toHash);
+                digest = md.digest();
+            } catch (NoSuchAlgorithmException e) {
+                Log.e(UTILS_TAG, e.getMessage());
+                //Shouldn't happen, if it happens: report to log
+            }
+            if (digest.length * 8 > numBits) {
+                byte[] trunk = new byte[(numBits % 8 > 0) ? numBits / 8 + 1 : numBits / 8];
+                System.arraycopy(digest, 0, trunk, 0, trunk.length);
+                digest = trunk;
+            }
+            return BitSet.valueOf(digest);
+        }
+        throw new IllegalArgumentException(NOT_VALID_NUMBIT_SHA_EXCEPTION_MSG);
+    }
+
+    /**
+     * Based on String.hashCode()
+     *
+     * @param key string of which generate hash code on 64 bits
+     * @return a 64 bits hash of the given String
+     */
+    private static long hash(@NonNull String key) {
+        long num = 0;
+        int n = key.length();
+        for (int i = 0; i < n; i++) {
+            num += key.charAt(i) * (63 ^ (n - (i + 1))); //original has 31 instead of 63
+        }
+        return num;
+    }
+
+    /**
+     * Method to compare two bitSets, useful to compare Distances
+     * compare(b1.xor(b2), b1.xor(b3)) ==> compare(D(b1, b2), D(b1, b3))
+     * ==> D(b1, b2) ? D(b1, b3)
+     * That is: is b1 closer to b2 or b3?
+     *
+     * @param lhs first bitSet
+     * @param rhs second bitSet
+     * @return a negative integer, zero, or a positive integer as rhs < lhs, rhs = lhs, rhs > lhs.
+     */
+    public static int compare(@NonNull BitSet lhs, @NonNull BitSet rhs) {
+        if (lhs.equals(rhs)) return 0;
+        BitSet distance = (BitSet) lhs.clone();
+        distance.xor(rhs);
+        int firstDifferent = distance.length() - 1;
+        if (firstDifferent == -1)
+            return 0; //redundant, won't be executed

```

Comment:
perché lo hai scritto se è ridondante?

Reply:
Perchè nella prima versione che ho scritto nel metodo non utilizzavo l'equals tra le due chiavi come primo test. e quindi se i bitset erano uguali si verificava con questo test. Alla fine l'ho lasciato come failsafe, però si potrebbe benissimo togliere. (Mi dispiaceva perché era carino)

Reply:
Originariamente si pensava di utilizzare direttamente i BitSet come chiavi, poi per varie ragioni (principalmente il fatto che non sono comparable) abbiamo aggiunto l'infrastruttura BinarySet (si potrebbe chiamare wrapper). 
Quindi questo metodo doveva servire per confrontare le distanze, adesso è quasi deprecato, si potrebbe pure tirare via, il fatto è che questa classe di utils e BinarySet potrebbero benissimo esistere in ambienti differenti, e questo metodo potrebbe risultare comodo.

Reply:
Chiarissimo 👍